lib/srv/desktop/rdp/rdpclient/README.md

```diff
@@ -82,3 +82,127 @@ Note that all memory allocated on the Go side must be freed on the Go side.
 Same goes for Rust. These languages have different memory management systems
 and can't free each others' memory. This is why you'll see the weird `free_*`
 hops from one side to the other.
+
+### Go/Rust Interface
+
+The `Client` struct in Rust is the object we pass back and forth between Rust and Go,
+so that Go can decide when to call its methods. We do this by allocating it on the heap
+in Rust and then leaking the memory into a raw pointer which is passed back to Go:
+
+```rust
+// Create the client
+let client = Client::new()
+// Move it to the heap
+let client = Box::new(client)
+// Convert it to a raw pointer, pass this back to Go.
+// Go is now responsible for the proper usage and
+// freeing of this memory.
+let client_ptr = Box::into_raw(client)
+```
+
+From that point on, the Client should be considered to be "owned" by Go (in Rust parlance),
+meaning that it's the Go side's responsibility to manage its memory. The two key elements
+of that responsibility are:
+
+1. Freeing the memory (once and only once) after it's done using it.
+2. Ensuring that the object isn't used after it's been freed.
+
+#### Enforcing proper memory semantics in Rust
+
+We call back into from Go using functions like
+
+```rust
+#[no_mangle]
+pub unsafe extern "C" fn call_into_rust_from_go(client_ptr: *const Client) -> CGOErrCode {
+  // Get immutable reference of Client
+  let client: &Client = client_ptr.as_ref().unwrap(); // unwrap for brevity, handle error in real code
+  // Do stuff with our immutable reference to Client
+  client.do_stuff()
+}
+```
+
+What's important here is that we stay disciplined in the following ways
+
+1. The `Client` itself is [`Sync`](https://doc.rust-lang.org/std/marker/trait.Sync.html)
+2. We access the client as an **immutable reference** (`&Client`).
+
+In order to understand why, its important to understand that when `call_into_rust_from_go`
+is called from Go, it can be:
+
+1. scheduled by the Go scheduler on any arbitrary thread, and
+2. can run concurrently with another thread also using the `Client`
+
+##### `Sync`
+
+With these constraints in mind, we can determine that `Client` must be forced to remain `Sync`,
+because the very definition of `Sync` "Types for which it is safe to share references between threads."
+Because the `Client` is owned by Go (meaning logically, we can only have references in Rust), and
+per 1 and 2 above the references in Rust must be shared between threads, the `Client` must be safe to
+do so (aka it must be `Sync`).
+
+We enforce `Sync` via this little Rust compiler hack
+
+```rust
+const _: () = {
+    const fn assert_sync<T: Sync>() {}
+    assert_sync::<Client>();
+};
+```
+
+##### Immutable Reference (`&Client`)
+
+One of the foundational memory guarantees of the Rust compiler is that you can only have
+a single mutable reference at a time. (In fact, if you have a mutable reference, you can't
+have an immutable reference to the same thing at the same time either).
```

Comment:
The actual names for `&mut T` and `&T` are "exclusive reference" (or "exclusive borrow") and "shared reference", which actually drives the point even further, IMO.

Reply:
ce1cb7b84f408a99697660ee212538e5ab694214